pandas/core/arrays/period.py

```diff
@@ -228,6 +303,183 @@ def start_time(self):
+            self.__class__.__name__,
+            [str(s) for s in self],
+            len(self),
+            self.dtype
+        )
+
+    def __len__(self):
+        return len(self._data)
+
+    def __setitem__(self, key, value):
+        if is_list_like(value):
+            if len(key) != len(value) and not com.is_bool_indexer(key):
+                msg = ("shape mismatch: value array of length '{}' does not "
+                       "match indexing result of length '{}'.")
+                raise ValueError(msg.format(len(key), len(value)))
+            if len(key) == 0:
+                return
+
+            value = period_array(value)
+
+            if self.freqstr != value.freqstr:
+                msg = DIFFERENT_FREQ_INDEX.format(self.freqstr, value.freqstr)
+                raise IncompatibleFrequency(msg)
+
+            value = value.asi8
+        elif isinstance(value, Period):
+
+            if self.freqstr != value.freqstr:
+                msg = DIFFERENT_FREQ_INDEX.format(self.freqstr, value.freqstr)
+                raise IncompatibleFrequency(msg)
+
+            value = value.ordinal
+        elif isna(value):
+            value = iNaT
+        else:
+            msg = ("'value' should be a 'Period', 'NaT', or array of those. "
+                   "Got '{}' instead.".format(type(value).__name__))
+            raise TypeError(msg)
+        self._data[key] = value
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        if allow_fill:
+            if isna(fill_value):
+                fill_value = iNaT
+            elif isinstance(fill_value, Period):
+                if self.freq != fill_value.freq:
+                    msg = DIFFERENT_FREQ_INDEX.format(
+                        self.freq.freqstr,
+                        fill_value.freqstr
+                    )
+                    raise IncompatibleFrequency(msg)
+
+                fill_value = fill_value.ordinal
+            else:
+                msg = "'fill_value' should be a Period. Got '{}'."
+                raise ValueError(msg.format(fill_value))
+
+        new_values = algos.take(self._data,
+                                indices,
+                                allow_fill=allow_fill,
+                                fill_value=fill_value)
+
+        return type(self)(new_values, self.freq)
+
+    def isna(self):
+        return self._data == iNaT
+
+    def fillna(self, value=None, method=None, limit=None):
+        # TODO(#20300)
+        # To avoid converting to object, we re-implement here with the changes
+        # 1. Passing `_ndarray_values` to func instead of self.astype(object)
+        # 2. Re-boxing output of 1.
+        # #20300 should let us do this kind of logic on ExtensionArray.fillna
+        # and we can use it.
+
+        if isinstance(value, ABCSeries):
+            value = value._values
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        mask = self.isna()
+
+        if is_array_like(value):
```

Comment:
this is only used on line 405  yes? if so maybe move this there? 

whey do we need to mask the value? I don't recall doing this anywhere else, e.g. shouldnt' setitem (on the PeriodArray) actually handle this?

Reply:
I think the implementation is correct.

If we didn't mask `value`, then we'd try to do something like

`array([0, 1, None])[False, False, True].__setitem__([1, 2, 3])`, which would fail since you're setting an array of length 3 into an array of length 1 (the mask original array).

Reply:
And it needs to be outside the `if mask.any()` so that we raise if they do something like


```
array([1, 2, 3]).fillna([1, 2])
```

i.e. filling with an array of the wrong length.

Reply:
that *should* fail, why are you allowing it?